From 513960989eac51cd6c3a2dd79d5d560890eba755 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Albrecht=20Dre=C3=9F?= <albrecht.dress@posteo.de>
Date: Sat, 25 Feb 2023 13:29:50 +0100
Subject: [PATCH] meson build: add option to enable Native Language Support
 (NLS)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Currently, NLS is never enabled when building Balsa using meson.  This
change adds the option to enable or disable NLS, and adds the proper
defines to config.h.

Details:
- meson_options.txt: define the `nls` option (default true)
- meson.build: iff nls is enabled, check for libintl.h and locale.h, and
  set the required #define items in config.h
- src/meson.build: building balsa_ab requires `GNOMELOCALEDIR` iff nls
  is enabled

Signed-off-by: Albrecht Dreß <albrecht.dress@posteo.de>
---
 meson.build       | 54 +++++++++++++++++++++++++++++------------------
 meson_options.txt |  5 +++++
 src/meson.build   |  4 +++-
 3 files changed, 42 insertions(+), 21 deletions(-)

diff --git a/meson.build b/meson.build
index 8a1cc78ac..8b1fffd95 100644
--- a/meson.build
+++ b/meson.build
@@ -66,6 +66,7 @@ libsecret     = get_option('libsecret')
 gcr           = get_option('gcr')
 more_warnings = get_option('more-warnings')
 help_files    = get_option('help-files')
+nls           = get_option('nls')
 
 # Documentation and Testing requirements for libnetclient
 libnetclient_docs = get_option('libnetclient-docs')
@@ -98,22 +99,6 @@ if help_files
   endif
 endif
 
-# _NL_MEASUREMENT_MEASUREMENT is an enum and not a define
-
-code = '''
-#include <langinfo.h>
-void func()
-{
-    char c;
-    c = *((unsigned char *) nl_langinfo(_NL_MEASUREMENT_MEASUREMENT));
-}
-'''
-result = compiler.compiles(code, name : '_NL_MEASUREMENT_MEASUREMENT test')
-if result
-  conf.set('HAVE__NL_MEASUREMENT_MEASUREMENT', 1,
-    description : 'Define if _NL_MEASUREMENT_MEASUREMENT is available')
-endif
-
 # #####################################################################
 # GNOME libs
 # #####################################################################
@@ -441,9 +425,38 @@ if compiler.has_function('ctime_r')
     description : 'Define to 1 if you have the ‘ctime_r’ function.')
 endif
 
-if compiler.has_header('locale.h')
-  conf.set('HAVE_LOCALE_H', 1,
-    description : 'Define to 1 if you have the <locale.h> header')
+#####################################################################
+# Native Language Support
+#####################################################################
+
+if nls
+  if not compiler.has_header('libintl.h')
+    error('libintl.h not found')
+  endif
+  if compiler.has_header('locale.h')
+    conf.set('HAVE_LOCALE_H', 1,
+      description : 'Define to 1 if you have the <locale.h> header')
+  endif
+  conf.set_quoted('GETTEXT_PACKAGE', meson.project_name(),
+    description : 'The `gettext` package name.')
+  conf.set('ENABLE_NLS', 1,
+    description : 'Define to 1 if Native Language Support is enabled.')
+
+  # _NL_MEASUREMENT_MEASUREMENT is an enum and not a define
+  code = '''
+  #include <langinfo.h>
+  char func(void)
+  {
+      char c;
+      c = *((unsigned char *) nl_langinfo(_NL_MEASUREMENT_MEASUREMENT));
+      return c;
+  }
+  '''
+  result = compiler.compiles(code, name : '_NL_MEASUREMENT_MEASUREMENT test')
+  if result
+    conf.set('HAVE__NL_MEASUREMENT_MEASUREMENT', 1,
+      description : 'Define if _NL_MEASUREMENT_MEASUREMENT is available')
+  endif
 endif
 
 # set more warnings.
@@ -595,6 +608,7 @@ summary = [
   '============= Final configuration =============',
   ' Installing into prefix: @0@'.format(get_option('prefix')),
   'Enable compile warnings: @0@'.format(more_warnings),
+  'Native Language Support: @0@'.format(nls),
   '            HTML widget: @0@'.format(html_widget),
   '              Use GNOME: @0@'.format(gnome_desktop),
   '           Use Canberra: @0@'.format(canberra),
diff --git a/meson_options.txt b/meson_options.txt
index 98859887e..bf599ac49 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -109,3 +109,8 @@ option('libnetclient-test',
   type        : 'boolean',
   value       : false,
   description : 'Check requirements for running libnetclient tests (see libnetclient/README; default=false)')
+
+option('nls',
+  type        : 'boolean',
+  value       : true,
+  description : 'use Native Language Support (default=true)')
diff --git a/src/meson.build b/src/meson.build
index 2623d86bb..00cba020c 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -116,8 +116,10 @@ balsa_ab_sources = [
   'address-book-config.c'
   ]
 
+balsa_ab_compile_arg = '-DGNOMELOCALEDIR="' + locale_dir + '"'
+balsa_ab_compile_dep = declare_dependency(compile_args : balsa_ab_compile_arg)
 executable('balsa-ab', balsa_ab_sources, balsa_ab_resources,
-           dependencies        : balsa_ab_deps,
+           dependencies        : balsa_ab_deps + [balsa_ab_compile_dep],
            include_directories : [top_include, libbalsa_include],
            link_with           : libbalsa_a,
            install             : true)
-- 
GitLab

